GHA build tweaks

wipe2238 opened the pull request:
- Rewritten to use single job again; Linux stuff (mostly) put into dedicated job step
  In my experience with GHA using separate job for each platform leads to more issues later, when/if build script grows. And it _will_ grow (64bit, mac, ...), so i'd like to address that early.

- Prepared for 64bit builds
  As source is not yet ready for that, they're currently commented.
  For reasons beyond my understanding GitHub does not provide a way to allow job to fail without failing whole build (which might lead to notifications spam when not-so-important-yet 64 builds obviously fail); while i am simulating it here, it's a bit of meh solution -- best can be done is force failed builds to be marked as successful :/
  Personally, i'd uncomment them when 64bit is on the table, but it's not a main dish _yet_. And once 64bit becomes a must-have, whole `can-fail` option should be removed.

- Added different build methods
  Added previous versions of GCC/VisualStudio
  Added CLang for a good measure, as it's sometimes able to catch nasties which fly through both GCC and/or VisualStudio without single warning.

- Uses hand-made matrix as builds configuration
  That should make adding different build methods easier, when/if needed.

---

Saw there's a work going for mac builds, which i'd be happy to include, but i have absolute zero experience with mac :(
If anyone can share details what would be key differences between linux and mac script (getting libs, artifact path, etc), i'll add it shortly.

burner1024 commented:
> Added CLang for a good measure, as it's sometimes able to catch nasties which fly through GCC and/or VisualStudio without single warning.

Maybe also add cppcheck?

wipe2238 commented:
> Maybe also add cppcheck?

Done, see top comment for details.

burner1024 commented:
easiest cppcheck option is just enable=all

wipe2238 commented:
Replaced Linux Clang with MacOS config in preparation for #24, added Ubuntu 20 fix from [check-ubuntu-20-04](../tree/check-ubuntu-20-04) branch, merged near all dynamic parts into single step.

alexbatalov commented:
I really think this is over engineered. Let's keep it for now. We'll review it once we have the game working on all major platforms.
